### ISSUE TITLE
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory

### ISSUE THREAD

**Issue by dennybritz** [appium/appium#13631]
## The problem

Same as https://github.com/appium/appium/issues/13321 and https://github.com/appium/appium/issues/13074 

Appium crashes after some time with `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory` due to memory leak. Process memory is continuously increasing.

I saw this problem with the latest stable version, then upgraded to `appium@beta` as suggested in the other issue. The problem is still the same.

## Environment

* Appium version (or git revision) that exhibits the issue: `1.16.0-beta.2`
* Last Appium version that did not exhibit the issue (if applicable): `N/A`
* Desktop OS/version used to run Appium: `macOS 10.15.1`
* Node.js version (unless using Appium.app|exe): `v13.1.0`
* Npm or Yarn package manager: `npm 6.13.1`
* Mobile platform/version under test: `iOS 12.4`
* Real device or emulator/simulator: `Real Device`
* Appium CLI or Appium.app|exe: `CLI`

## Details

The application is taking screenshots with `get_screenshot_as_file` in a loop, same as the other issue(s) linked above. Here's a heap snapshot taken a 10 minutes into the run with memory usage already at several GB:

[Heap.20191125.115127.20733.0.006.heapsnapshot.zip](https://github.com/appium/appium/files/3884647/Heap.20191125.115127.20733.0.006.heapsnapshot.zip)

I took a brief look at the dump and it seems like all screenshots are left in memory?

## Link to Appium logs

https://gist.github.com/dennybritz/c727f4ad03379340713fc3d3a19b0dc5

**Comment by mykola-mokhnach** [comment]
@dennybritz Can you please apply the patch https://github.com/appium/appium-base-driver/pull/370 to your local sources and try again?

**Comment by dennybritz** [comment]
I applied the changes to `/usr/local/lib/node_modules/appium/node_modules/appium-base-driver` but I don't think that fixed it, the memory usage still keeps increasing.

**Comment by mykola-mokhnach** [comment]
Can you provide the updated snapshot?

**Comment by dennybritz** [comment]
Here's a snapshot: 
[Heap.20191125.225624.40557.0.002.heapsnapshot.zip](https://github.com/appium/appium/files/3886911/Heap.20191125.225624.40557.0.002.heapsnapshot.zip)

**Comment by mykola-mokhnach** [comment]
The recent snapshot seems to be the same.

Make sure the appium-base-driver module is deduplicated. It is necessary to make sure the instance is used by both the umbrella driver (the main Appium module)  and appium-xcuitest-driver. https://github.com/appium/appium/issues/13321#issuecomment-540114647 contains more details on how to deploy the changes locally

**Comment by dennybritz** [comment]
I haven't been able to test this. I don't know why patching the module didn't work, but I tried building Appium from scratch while linking in the appium-base-driver module, but when doing that, I get the following error when trying to execute commands:

```
[debug] [W3C (36f3e66a)] Encountered internal error running command: NoSuchDriverError: A session is either terminated or not started
[debug] [W3C (36f3e66a)]     at asyncHandler (/Users/dennybritz/github/appium-base-driver/lib/protocol/protocol.js:252:15)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/lib/protocol/protocol.js:432:15
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:95:5)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/route.js:137:13)
[debug] [W3C (36f3e66a)]     at Route.dispatch (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/route.js:112:3)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:95:5)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:281:22
[debug] [W3C (36f3e66a)]     at param (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:354:14)
[debug] [W3C (36f3e66a)]     at param (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:365:14)
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:410:3)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at logger (/Users/dennybritz/github/appium-base-driver/node_modules/morgan/index.js:144:5)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:95:5)
[debug] [W3C (36f3e66a)]     at trim_prefix (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:317:13)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:284:7
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:335:12)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at jsonParser (/Users/dennybritz/github/appium-base-driver/node_modules/body-parser/lib/types/json.js:110:7)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:95:5)
[debug] [W3C (36f3e66a)]     at trim_prefix (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:317:13)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:284:7
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:335:12)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:91:12)
[debug] [W3C (36f3e66a)]     at trim_prefix (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:317:13)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:284:7
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:335:12)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:91:12)
[debug] [W3C (36f3e66a)]     at trim_prefix (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:317:13)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:284:7
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:335:12)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at methodOverride (/Users/dennybritz/github/appium-base-driver/node_modules/method-override/index.js:65:14)
[HTTP] <-- GET /wd/hub/session/36f3e66a-2f3f-446d-8229-08056ee2aacc/screenshot 404 24 ms - 4039
[HTTP] 
[HTTP] --> GET /wd/hub/session/36f3e66a-2f3f-446d-8229-08056ee2aacc/screenshot
[HTTP] {}
[debug] [W3C (36f3e66a)] Encountered internal error running command: NoSuchDriverError: A session is either terminated or not started
[debug] [W3C (36f3e66a)]     at asyncHandler (/Users/dennybritz/github/appium-base-driver/lib/protocol/protocol.js:252:15)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/lib/protocol/protocol.js:432:15
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:95:5)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/route.js:137:13)
[debug] [W3C (36f3e66a)]     at Route.dispatch (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/route.js:112:3)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:95:5)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:281:22
[debug] [W3C (36f3e66a)]     at param (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:354:14)
[debug] [W3C (36f3e66a)]     at param (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:365:14)
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:410:3)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at logger (/Users/dennybritz/github/appium-base-driver/node_modules/morgan/index.js:144:5)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:95:5)
[debug] [W3C (36f3e66a)]     at trim_prefix (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:317:13)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:284:7
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:335:12)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at jsonParser (/Users/dennybritz/github/appium-base-driver/node_modules/body-parser/lib/types/json.js:110:7)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:95:5)
[debug] [W3C (36f3e66a)]     at trim_prefix (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:317:13)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:284:7
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:335:12)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:91:12)
[debug] [W3C (36f3e66a)]     at trim_prefix (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:317:13)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:284:7
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:335:12)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at Layer.handle [as handle_request] (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/layer.js:91:12)
[debug] [W3C (36f3e66a)]     at trim_prefix (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:317:13)
[debug] [W3C (36f3e66a)]     at /Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:284:7
[debug] [W3C (36f3e66a)]     at Function.process_params (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:335:12)
[debug] [W3C (36f3e66a)]     at next (/Users/dennybritz/github/appium-base-driver/node_modules/express/lib/router/index.js:275:10)
[debug] [W3C (36f3e66a)]     at methodOverride (/Users/dennybritz/github/appium-base-driver/node_modules/method-override/index.js:65:14)
```

**Comment by mykola-mokhnach** [comment]
The log above seems odd. I'll spend some time to check the stuff locally, although I cannot say how long it would take

**Comment by dennybritz** [comment]
For what it's worth, I did built the `1.16.0-beta.2` version locally and it worked fine, but once I link in the updated `appium-base-driver module` I get the above and it immediately closes the session.

**Comment by mykola-mokhnach** [comment]
Try to also apply the changes from https://github.com/appium/appium/pull/13635

This should help

**Comment by dennybritz** [comment]
That seems to have fixed the session close issue but the memory leak problem is still there:

[Heap.20191126.165021.62039.0.003.heapsnapshot.zip](https://github.com/appium/appium/files/3890614/Heap.20191126.165021.62039.0.003.heapsnapshot.zip)

**Comment by mykola-mokhnach** [comment]
Thanks for your patience. We really appreciate your support. Please try if there are still memory leaks after https://github.com/appium/appium-base-driver/pull/371 is applied

**Comment by dennybritz** [comment]
Yep, this seems to fix it!  🎉

**Comment by mykola-mokhnach** [comment]
Perfect. We'll merge the changes by the end of the day. Although, it will take some time until new versions of all components will be published (I'm not sure if Appium 1.16 is going to include this change)

CC @dpgraham

**Comment by mykola-mokhnach** [comment]
The fix has been published to beta
